runners/flink/src/test/java/org/apache/beam/runners/flink/PortableExecutionTest.java

```diff
@@ -0,0 +1,234 @@
+          return (container) -> RemoteEnvironment.forHandler(container, controlClient);
+        } catch (Exception ex) {
+          throw new RuntimeException(ex);
+        }
+      }
+    };
+  }
+
+  @Before
+  public void setup() {
+    DockerJobBundleFactory.factory = this::createJobBundleFactory;
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    sdkHarnessExecutor.shutdownNow();
+    try {
+      sdkHarnessExecutorFuture.get();
+    } catch (ExecutionException e) {
+      if (e.getCause() instanceof RuntimeException
+          && e.getCause().getCause() instanceof InterruptedException) {
+        // expected
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  private static ArrayList<KV<String, Iterable<Long>>> outputValues = new ArrayList<>();
+
+  @Test
+  public void testExecution() throws Exception {
+    Pipeline p = Pipeline.create();
+    p.apply("impulse", Impulse.create())
+        .apply(
+            "create",
+            ParDo.of(
+                new DoFn<byte[], String>() {
+                  @ProcessElement
+                  public void process(ProcessContext ctxt) {
+                    ctxt.output("zero");
+                    ctxt.output("one");
+                    ctxt.output("two");
+                  }
+                }))
+        .apply(
+            "len",
+            ParDo.of(
+                new DoFn<String, Long>() {
+                  @ProcessElement
+                  public void process(ProcessContext ctxt) {
+                    ctxt.output((long) ctxt.element().length());
+                  }
+                }))
+        .apply("addKeys", WithKeys.of("foo"))
+        // Use some unknown coders
+        .setCoder(KvCoder.of(StringUtf8Coder.of(), BigEndianLongCoder.of()))
+        // Force the output to be materialized
+        .apply("gbk", GroupByKey.create())
+        .apply(
+            "collect",
+            ParDo.of(
+                new DoFn<KV<String, Iterable<Long>>, Void>() {
+                  @ProcessElement
+                  public void process(ProcessContext ctx) {
+                    outputValues.add(ctx.element());
+                  }
+                }));
+
+    RunnerApi.Pipeline pipelineProto = PipelineTranslation.toProto(p);
+
+    outputValues.clear();
+    // execute the pipeline
+    ListeningExecutorService executor =
+        MoreExecutors.listeningDecorator(Executors.newCachedThreadPool());
+    FlinkPipelineOptions options = PipelineOptionsFactory.as(FlinkPipelineOptions.class);
+    options.setFlinkMaster("[local]");
+    options.setStreaming(isStreaming);
+    FlinkJobInvocation jobInvocation =
+        FlinkJobInvocation.create(
+            "fakeId",
+            "fakeRetrievalToken",
+            executor,
+            pipelineProto,
+            options,
+            Collections.EMPTY_LIST);
+    jobInvocation.start();
+    long timeout = System.currentTimeMillis() + 30 * 1000;
```

Comment:
30 sec timeout for a unit test seems to be too long. Shall we reduce it or else we can make it an integration test.

Reply:
The test completes in ~9 seconds for both batch and streaming on my macbook. It would only take 30s in the case of a failure..